clients/include/cblas_interface.hpp

```diff
@@ -95,6 +95,7 @@ inline void cblas_gemm(hipblasOperation_t       transA,
 {
     switch(tc)
     {
+    case HIP_R_16F: // setting compute_type to f16_r will fallback to f32_r
```

Comment:
Client side doesn't know if the hipblaslt library go f16 compute or fallback f32 compute.
I would suggest letting the host-side calculation be based on the original user-desired compute type

Reply:
Do you mean to move the fallback logic inside the `testing_matmul_with_bias` function?
https://github.com/ROCm/hipBLASLt/blob/576d0c76701c32c74686f0a2394d42bb2c6ef469/clients/include/testing_matmul.hpp#L2587-L2592
Otherwise, it'll result in a norm and unit error when comparing the f32-computed GEMM from the device with the f16-computed GEMM from the host. The client side must calculate GEMM using the fallback f32 compute type.

Reply:
have you tried and see the norm error (or unit error ) out of expectation tolerance indeed?

Reply:
> ```yaml
> hipBLASLt version: 1200
> 
> Query device success: there are 1 devices
> -------------------------------------------------------------------------------
> Device ID 0 : AMD Radeon Graphics gfx1201
> with 17.1 GB memory, max. SCLK 2400 MHz, max. MCLK 1258 MHz, compute capability 12.0
> maxGridDimX 2147483647, sharedMemPerBlock 65.5 KB, maxThreadsPerBlock 1024, warpSize 32
> -------------------------------------------------------------------------------
> info: parsing of test data may take a couple minutes before any test output appears...
> 
> /home/haosheng/projects/hipBLASLt+add_support_for_fallback_from_compute_type_f16_to_f32/build_576d0c/debug/clients/staging/hipblaslt_gentest.py --template /home/haosheng/projects/hipBLASLt+add_support_for_fallback_from_compute_type_f16_to_f32/build_576d0c/debug/clients/staging/hipblaslt_template.yaml -o /tmp/hipblaslt-iCJqyd ../../../../.vscode/bench.yaml
> Note: Google Test filter = *quick*
> [==========] Running 1 test from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 1 test from _/matmul_test
> [ RUN      ] _/matmul_test.matmul/quick_matmul_skinny_real_precisions_1b_f16_rf16_rf16_rf16_rf16_r_TT_8_8_8_1_8_8_0_8_8_1
> /home/haosheng/projects/hipBLASLt+add_support_for_fallback_from_compute_type_f16_to_f32/clients/gtest/../include/near.hpp:244: Failure
> The difference between double(hCPU[i + j * size_t(lda) + k * strideA]) and double(hGPU[i + j * size_t(lda) + k * strideA]) is 1, which exceeds abs_error, where
> double(hCPU[i + j * size_t(lda) + k * strideA]) evaluates to 1,
> double(hGPU[i + j * size_t(lda) + k * strideA]) evaluates to 0, and
> abs_error evaluates to 0.080000000000000002.
> 
> [  FAILED  ] _/matmul_test.matmul/quick_matmul_skinny_real_precisions_1b_f16_rf16_rf16_rf16_rf16_r_TT_8_8_8_1_8_8_0_8_8_1, where GetParam() = { function: "matmul", name: "matmul_skinny_real_precisions_1b", category: "quick", known_bug_platforms: "", alpha: 1, beta: 0, stride_a: 0x104c288, stride_b: 0x104c388, stride_c: 0x104c488, stride_d: 0x104c588, stride_e: 0x104c688, user_allocated_workspace: 134217728, M: 0x104c790, N: 0x104c890, K: 0x104c990, lda: 0x104ca90, ldb: 0x104cb90, ldc: 0x104cc90, ldd: 0x104cd90, lde: 0x104ce90, batch_count: 1, iters: 10, cold_iters: 2, algo: 0, solution_index: -1, requested_solution_num: 1, a_type: f16_r, b_type: f16_r, c_type: f16_r, d_type: f16_r, compute_type: f16_r, compute_input_typeA: non-supported type, compute_input_typeB: non-supported type, scale_type: f32_r, initialization: "rand_int", gpu_arch: "", pad: 4096, grouped_gemm: 0, threads: 0, streams: 0, devices:  (164 ms)
> [----------] 1 test from _/matmul_test (164 ms total)
> 
> [----------] Global test environment tear-down
> [==========] 1 test from 1 test suite ran. (165 ms total)
> [  PASSED  ] 0 tests.
> [  FAILED  ] 1 test, listed below:
> [  FAILED  ] _/matmul_test.matmul/quick_matmul_skinny_real_precisions_1b_f16_rf16_rf16_rf16_rf16_r_TT_8_8_8_1_8_8_0_8_8_1, where GetParam() = { function: "matmul", name: "matmul_skinny_real_precisions_1b", category: "quick", known_bug_platforms: "", alpha: 1, beta: 0, stride_a: 0x104c288, stride_b: 0x104c388, stride_c: 0x104c488, stride_d: 0x104c588, stride_e: 0x104c688, user_allocated_workspace: 134217728, M: 0x104c790, N: 0x104c890, K: 0x104c990, lda: 0x104ca90, ldb: 0x104cb90, ldc: 0x104cc90, ldd: 0x104cd90, lde: 0x104ce90, batch_count: 1, iters: 10, cold_iters: 2, algo: 0, solution_index: -1, requested_solution_num: 1, a_type: f16_r, b_type: f16_r, c_type: f16_r, d_type: f16_r, compute_type: f16_r, compute_input_typeA: non-supported type, compute_input_typeB: non-supported type, scale_type: f32_r, initialization: "rand_int", gpu_arch: "", pad: 4096, grouped_gemm: 0, threads: 0, streams: 0, devices: 
> 
>  1 FAILED TEST
> hipBLASLt version: 1200
> 
> command line: ../staging/hipblaslt-test --yaml ../../../../.vscode/bench.yaml --gtest_filter=*quick* 
> ```
> 
> ```diff
> diff --git a/clients/common/cblas_interface.cpp b/clients/common/cblas_interface.cpp
> index 663f6c9c..8d016c6a 100644
> --- a/clients/common/cblas_interface.cpp
> +++ b/clients/common/cblas_interface.cpp
> @@ -324,17 +324,17 @@ void cast_mul(customVector<TcCast>& dst,
>                                              k,
>                                              size);
>          break;
> -    case HIP_R_16BF:
> -        cast_mul<TcCast, Tc, hip_bfloat16>(dst,
> -                                           static_cast<const hip_bfloat16*>(src),
> -                                           isScaleAVec,
> -                                           scaleAVec,
> -                                           AlphaVec,
> -                                           transA,
> -                                           m,
> -                                           k,
> -                                           size);
> -        break;
> +    // case HIP_R_16BF:
> +        // cast_mul<TcCast, Tc, hip_bfloat16>(dst,
> +        //                                    static_cast<const hip_bfloat16*>(src),
> +        //                                    isScaleAVec,
> +        //                                    scaleAVec,
> +        //                                    AlphaVec,
> +        //                                    transA,
> +        //                                    m,
> +        //                                    k,
> +        //                                    size);
> +        // break;
>      case HIP_R_8F_E4M3_FNUZ:
>          cast_mul<TcCast, Tc, hipblaslt_f8_fnuz>(dst,
>                                                  static_cast<const hipblaslt_f8_fnuz*>(src),
> @@ -532,17 +532,17 @@ void cast_mul_with_Tci(customVector<TcCast>& dst,
>              k,
>              size);
>          break;
> -    case HIP_R_16BF:
> -        cast_mul_with_Tci<TcCast, Tc, TciACast, hip_bfloat16>(dst,
> -                                                              static_cast<const hip_bfloat16*>(src),
> -                                                              isScaleAVec,
> -                                                              scaleAVec,
> -                                                              AlphaVec,
> -                                                              transA,
> -                                                              m,
> -                                                              k,
> -                                                              size);
> -        break;
> +    // case HIP_R_16BF:
> +    //     cast_mul_with_Tci<TcCast, Tc, TciACast, hip_bfloat16>(dst,
> +    //                                                           static_cast<const hip_bfloat16*>(src),
> +    //                                                           isScaleAVec,
> +    //                                                           scaleAVec,
> +    //                                                           AlphaVec,
> +    //                                                           transA,
> +    //                                                           m,
> +    //                                                           k,
> +    //                                                           size);
> +    //     break;
>      case HIP_R_8F_E4M3_FNUZ:
>          cast_mul_with_Tci<TcCast, Tc, TciACast, hipblaslt_f8_fnuz>(
>              dst,
> @@ -649,10 +649,10 @@ void cast_mul_with_Tci(customVector<TcCast>& dst,
>          cast_mul_with_Tci<TcCast, Tc, hipblasLtHalf>(
>              dst, src, TiA, isScaleAVec, scaleAVec, AlphaVec, transA, m, k, size);
>          break;
> -    case HIP_R_16BF:
> -        cast_mul_with_Tci<TcCast, Tc, hip_bfloat16>(
> -            dst, src, TiA, isScaleAVec, scaleAVec, AlphaVec, transA, m, k, size);
> -        break;
> +    // case HIP_R_16BF:
> +    //     cast_mul_with_Tci<TcCast, Tc, hip_bfloat16>(
> +    //         dst, src, TiA, isScaleAVec, scaleAVec, AlphaVec, transA, m, k, size);
> +    //     break;
>      case HIP_R_8F_E4M3_FNUZ:
>          cast_mul_with_Tci<TcCast, Tc, hipblaslt_f8_fnuz>(
>              dst, src, TiA, isScaleAVec, scaleAVec, AlphaVec, transA, m, k, size);
> @@ -1056,6 +1056,7 @@ void cblas_gemm(hipblasOperation_t       transA,
>                                   hipDataType              TciB,        \
>                                   bool                     alt);
>  
> +CREATEFUNCTION(hipblasLtHalf)
>  CREATEFUNCTION(float)
>  CREATEFUNCTION(double)
>  CREATEFUNCTION(int32_t)
> diff --git a/clients/include/cblas_interface.hpp b/clients/include/cblas_interface.hpp
> index 1a05ca76..162872f7 100644
> --- a/clients/include/cblas_interface.hpp
> +++ b/clients/include/cblas_interface.hpp
> @@ -96,6 +96,32 @@ inline void cblas_gemm(hipblasOperation_t       transA,
>      switch(tc)
>      {
>      case HIP_R_16F: // setting compute_type to f16_r will fallback to f32_r
> +        cblas_gemm<hipblasLtHalf>(transA,
> +                          transB,
> +                          m,
> +                          n,
> +                          k,
> +                          alpha.f16,
> +                          A,
> +                          lda,
> +                          B,
> +                          ldb,
> +                          beta.f16,
> +                          C,
> +                          ldc,
> +                          (const hipblasLtHalf*)AlphaVec,
> +                          (const hipblasLtHalf*)scaleA,
> +                          (const hipblasLtHalf*)scaleB,
> +                          *(hipblasLtHalf*)scaleD,
> +                          isScaleAVec,
> +                          isScaleBVec,
> +                          tiA,
> +                          tiB,
> +                          to,
> +                          tc,
> +                          tciA,
> +                          tciB,
> +                          alt);
>      case HIP_R_32F:
>          cblas_gemm<float>(transA,
>                            transB,
> @@ -119,7 +145,7 @@ inline void cblas_gemm(hipblasOperation_t       transA,
>                            tiA,
>                            tiB,
>                            to,
> -                          HIP_R_32F,
> +                          tc,
>                            tciA,
>                            tciB,
>                            alt);
> ```
> 
> commented out bf16 because compiler was showing below error:
> 
> ```yaml
> clients/common/cblas_interface.cpp:261:64: error: use of overloaded operator '*' is ambiguous (with operand types 'const hip_bfloat16' and '_Float16')
>   261 |                         dst[i]      = static_cast<TcCast>(A[i] * scaleA);
> ```

It doesn't make any sense that has abs error for rand_init testing for compute fp16.

Reply:
The attached modification is missing the cblas_hgemm calculation, resulting in an incorrect result. Currently, we fallback to f32 on CPU for calculations. If we want to perform calculations with f16 precision in the future, we'll need to implement `hgemm` on the CPU, as LAPACK does not support `hgemm`.